README.md
# bmp180sensor

Node.js I2C driver for the BMP180 pressure and temperature sensor on
Linux boards like the Raspberry Pi or BeagleBone.

Supports Node.js versions 11+

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## v0.0.1 - 2020-05-15

### Commits

- Initial commit [`4764b4d`](https://github.com/lushonline/bmp180sensor/commit/4764b4d5605b879ae774c45158e2f91be70dcd93)


test/bmp180sensor-test.js
const chai = require('chai');
const { describe, it } = require('mocha');
const { openSync } = require('i2c-bus');

const { Bmp180sensor } = require('../src/bmp180sensor.js');
const { oversamplingSettings } = require('../src/bmp180sensor.constants');

// This is the start of the set of tests associated with the test cases
describe('Bmp180sensor Test Suite', () => {
  it('HELPERS1: Confirm error thrown if we dont pass I2c bus', () => {
    const badFn = () => {
      // eslint-disable-next-line no-unused-vars
      const test = new Bmp180sensor();
    };
    chai.expect(badFn).to.throw(TypeError, 'Invalid bus it must a number or an i2c-bus.Bus');
  });

  it('HELPERS2: Confirm error thrown if we pass I2c a none numeric bus', () => {
    const badFn = () => {
      // eslint-disable-next-line no-unused-vars
      const test = new Bmp180sensor('bad');
    };
    chai.expect(badFn).to.throw(TypeError, 'Invalid bus it must a number or an i2c-bus.Bus');
  });

  it('HELPERS3: Confirm error thrown if we pass I2c a null bus', () => {
    const badFn = () => {
      // eslint-disable-next-line no-unused-vars
      const test = new Bmp180sensor(null);
    };
    chai.expect(badFn).to.throw(TypeError, 'Invalid bus it must a number or an i2c-bus.Bus');
  });

  it('HELPERS4: Confirm we can pass i2c-bus Bus object', () => {
    const i2cbus = openSync(1);
    const bmp180 = new Bmp180sensor(i2cbus);
    chai.expect(bmp180.isValidBus).to.equal(true);
  });

  it('HELPERS5: Confirm we can pass numeric bus', () => {
    const bmp180 = new Bmp180sensor(1);
    chai.expect(bmp180.isValidBus).to.equal(true);
  });

  it('COMMANDS1: name() - Confirm we can retrieve our sensor name', () => {
    const bmp180 = new Bmp180sensor(1);
    chai.expect(bmp180.name).to.equal('BMP180');
  });

  it('COMMANDS2: isValidChipId() - need to have BMP180 connected to bus 1, at address 0x77', async () => {
    const bmp180 = new Bmp180sensor(1);
    let success = false;

    try {
      success = await bmp180.isValidChipId();
    } catch (e) {
      throw new Error(`Got an unexpected exception: ${e.message}`);
    }

    chai.expect(success).to.equal(true);
  });

  it('COMMANDS3: softreset() - need to have BMP180 connected to bus 1, at address 0x77', async () => {
    const bmp180 = new Bmp180sensor(1);
    let success = false;

    try {
      success = await bmp180.softreset();
    } catch (e) {
      throw new Error(`Got an unexpected exception: ${e.message}`);
    }

    chai.expect(success).to.equal(true);
  });

  it('COMMANDS4: calibrate() - need to have BMP180 connected to bus 1, at address 0x77', async () => {
    const bmp180 = new Bmp180sensor(1);
    let success = false;

    try {
      success = await bmp180.calibrate();
    } catch (e) {
      throw new Error(`Got an unexpected exception: ${e.message}`);
    }

    chai.expect(success).to.equal(true);
    chai.expect(bmp180.isCalibrated).to.equal(true);
  });

  it('COMMANDS5: initialize() - need to have BMP180 connected to bus 1, at address 0x77', async () => {
    const bmp180 = new Bmp180sensor(1);
    let success = false;

    try {
      success = await bmp180.initialize();
    } catch (e) {
      throw new Error(`Got an unexpected exception: ${e.message}`);
    }

    chai.expect(success).to.equal(true);
    chai.expect(bmp180.isCalibrated).to.equal(true);
    chai.expect(bmp180.isInitialized).to.equal(true);
  });

  it('COMMANDS6: temperature() - need to have BMP180 connected to bus 1, at address 0x77', async () => {
    const bmp180 = new Bmp180sensor(1);
    let value = null;

    try {
      await bmp180.initialize().then(async () => {
        value = await bmp180.temperature();
      });
    } catch (e) {
      throw new Error(`Got an unexpected exception: ${e.message}`);
    }

    chai.expect(value).to.be.a('number');
    chai.expect(bmp180.isCalibrated).to.equal(true);
    chai.expect(bmp180.isInitialized).to.equal(true);
  });

  it('COMMANDS7: pressure() - need to have BMP180 connected to bus 1, at address 0x77', async () => {
    const bmp180 = new Bmp180sensor(1);
    let value = null;

    try {
      await bmp180.initialize().then(async () => {
        value = await bmp180.pressure();
      });
    } catch (e) {
      throw new Error(`Got an unexpected exception: ${e.message}`);
    }

    chai.expect(value).to.be.a('number');
    chai.expect(bmp180.isCalibrated).to.equal(true);
    chai.expect(bmp180.isInitialized).to.equal(true);
  });

  it('COMMANDS8: close() bus created by sensor class - need to have BMP180 connected to bus 1, at address 0x77', async () => {
    const bmp180 = new Bmp180sensor(1);
    let success = false;

    try {
      await bmp180.initialize().then(async () => {
        success = await bmp180.close();
      });
    } catch (e) {
      throw new Error(`Got an unexpected exception: ${e.message}`);
    }

    chai.expect(success).to.equal(true);
  });

  it('COMMANDS9: close() bus created and passed to sensor class - need to have BMP180 connected to bus 1, at address 0x77', async () => {
    const i2cbus = openSync(1);
    const bmp180 = new Bmp180sensor(i2cbus);
    let success = null;

    try {
      await bmp180.initialize().then(async () => {
        success = await bmp180.close();
      });
    } catch (e) {
      throw new Error(`Got an unexpected exception: ${e.message}`);
    }

    chai.expect(success).to.equal(false);
  });

  it('OSS1: Confirm we can get the pressure oversampling', () => {
    const bmp180 = new Bmp180sensor(1);
    const value = bmp180.oss;
    chai.expect(value).to.deep.equal(oversamplingSettings.standard);
  });

  it('OSS2: Confirm we can set the pressure oversampling', () => {
    const bmp180 = new Bmp180sensor(1);
    const originalvalue = bmp180.oss;

    bmp180.oss = oversamplingSettings.high;
    const value = bmp180.oss;

    chai.expect(originalvalue).to.deep.equal(oversamplingSettings.standard);
    chai.expect(value).to.deep.equal(oversamplingSettings.high);
  });
});


src/lib/i2csensor.js
const { openSync } = require('i2c-bus');
const debug = require('debug')('I2csensor');
const PromiseRetry = require('promise-retry');
const _ = require('lodash');

const { i2csensorDefaults } = require('./i2csensor.defaults');

/**
 * I2csensor Base Class
 *
 * @class I2csensor
 */
class I2csensor {
  /**
   *Creates an instance of I2csensor.
   * @param {string} name The name for the Sensor
   * @param {object|number} bus An i2c-bus Bus class or the number for the bus
   * @param {number} address The I2C address
   * @param {i2csensorDefaults} options The configuration options
   *
   * @memberof I2csensor
   */
  constructor(name, bus, address, options) {
    this._name = name;
    this._address = this._validatedAddress(address);

    this._bus = null;
    this._busnumber = null;

    this._validbus = false;

    switch (typeof bus) {
      case 'object':
        this._busnumber = null;
        this._bus = this._validatedBus(bus).promisifiedBus();
        this._validbus = true;
        break;
      case 'number':
        this._busnumber = this._validatedBusnumber(bus);
        this._bus = openSync(this._busnumber).promisifiedBus();
        this._validbus = true;
        break;
      default:
        // code block
        this._validbus = false;
        this._busnumber = null;
        this._bus = null;
        throw new TypeError('Invalid bus it must a number or an i2c-bus.Bus');
    }

    this._opts = _.merge({}, i2csensorDefaults, options);
  }

  get name() {
    return this._name;
  }

  /**
   * Get the current chip address to use, return as Base 10
   *
   * @memberof I2csensor
   */
  get address() {
    return this._address;
  }

  /**
   * Set the current chip address to use
   *
   * @memberof I2csensor
   */
  set address(value) {
    this._address = this._validatedAddress(value);
  }

  /**
   * Confirm a Valid I2C Bus has been supplied
   *
   * @readonly
   * @memberof I2csensor
   */
  get isValidBus() {
    return this._validbus;
  }

  /**
   * Confirm a Valid bus parameter
   *
   * @param {*} bus Validates the bus supplied is valid
   * @throws Will throw an error if invalid.
   * @returns {object}
   *
   * @memberof I2csensor
   */
  _validatedBus = (bus) => {
    if (_.isNil(bus)) {
      throw new TypeError('Invalid bus it must a number or an i2c-bus.Bus');
    }
    if (bus.constructor.name !== 'Bus') {
      throw new TypeError('Invalid bus it must a number or an i2c-bus.Bus');
    }
    return bus;
  };

  /**
   * Confirm a Valid bus number parameter
   *
   * @param {*} busnumber Validates the bus number is valid
   * @throws Will throw an error if invalid.
   * @returns {number}
   *
   * @memberof I2csensor
   */
  _validatedBusnumber = (busnumber) => {
    if (!Number.isInteger(busnumber) || busnumber < 0) {
      throw new TypeError(`Invalid I2C bus number ${busnumber}`);
    }
    return busnumber;
  };

  /**
   * Confirm a Valid I2C address
   *
   * @param {*} address Validates the address is valid
   * @throws Will throw an error if invalid.
   * @returns {number}
   *
   * @memberof I2csensor
   */
  _validatedAddress = (address) => {
    if (!Number.isInteger(address) || address < 0 || address > 0x7f) {
      throw new TypeError(`Invalid I2C address ${address}`);
    }
    return address;
  };

  /**
   * Confirm a Valid I2C command/register
   *
   * @param {*} command Validates the command is valid
   * @throws Will throw an error if invalid.
   * @returns {number}
   *
   * @memberof I2csensor
   */
  _validatedCommand = (command) => {
    if (!Number.isInteger(command) || command < 0 || command > 0xff) {
      throw new TypeError(`Invalid I2C command ${command}`);
    }
    return command;
  };

  /**
   * Confirm a Valid Buffer
   *
   * @param {*} buffer Validates the buffer is a valid buffer
   * @throws Will throw an error if invalid.
   * @returns {Buffer}
   *
   * @memberof I2csensor
   */
  _validatedBuffer = (buffer) => {
    if (!Buffer.isBuffer(buffer)) {
      throw new TypeError(`Invalid buffer ${buffer}`);
    }
    return buffer;
  };

  /**
   * Read a byte from the specified command/register
   *
   * @param {integer} command The command/register to read from
   * @throws Will throw an error if invalid command/register
   * @returns {Promise<number>} Promise that will be resolved with an unsigned integer (byte) in the range 0 to 255 on success.
   * @memberof I2csensor
   */
  _readI2cByte(command) {
    const cmd = this._validatedCommand(command);
    debug('readI2cByte. address: %h command: %h', this._address, cmd);
    return PromiseRetry((retry, number) => {
      debug('readBreadI2cByteyte. attempt: %d', number);
      return this._bus.readByte(this._address, cmd).catch((err) => {
        debug('readI2cByte. error: %o', err);
        retry();
      });
    }, this._opts.retryOptions);
  }

  /**
   * Write a byte to the specified command/register
   *
   * @param {integer} command The command/register to write to
   * @param {integer} byte The value to write
   * @returns {Promise} Promise that will be resolved with no arguments on success.
   * @throws Will throw an error if invalid command/register
   * @memberof I2csensor
   */
  _writeI2cByte(command, byte) {
    const cmd = this._validatedCommand(command);
    debug('writeI2cByte. address: %h command: %h byte: %h', this._address, cmd, byte);
    return PromiseRetry((retry, number) => {
      debug('writeI2cByte. attempt: %d', number);
      return this._bus.writeByte(this._address, cmd, byte).catch((err) => {
        debug('writeI2cByte. error: %o', err);
        retry();
      });
    }, this._opts.retryOptions);
  }

  /**
   * Read a buffer from the specified command/register
   *
   * @param {integer} command The command/register to read from
   * @param {Buffer} buffer The buffer to read
   * @throws Will throw an error if invalid command/register
   * @throws Will throw an error if invalid buffer
   * @returns {Promise<{bytesRead: number, buffer: Buffer}>} Returns a Promise that on success will be resolved with an object with a bytesRead property identifying the number of bytes read, and a buffer property contiining the read block.
   */
  _readI2cBlock(command, buffer) {
    const cmd = this._validatedCommand(command);
    const buff = this._validatedBuffer(buffer);
    const { length } = buff;
    debug('readI2cBlock. address: %h command: %h length: %h', this._address, cmd, length);
    return PromiseRetry((retry, number) => {
      debug('readI2cBlock. attempt: %d', number);
      return this._bus.readI2cBlock(this._address, cmd, length, buff).catch((err) => {
        debug('readI2cBlock. error: %o', err);
        retry();
      });
    }, this._opts.retryOptions);
  }

  /**
   * Write a buffer to the specified command/register
   *
   * @param {integer} command The command/register to write to
   * @param {Buffer} buffer The buffer to write
   * @throws Will throw an error if invalid command/register
   * @throws Will throw an error if invalid buffer
   * @returns {Promise<{bytesWritten: number, buffer: Buffer}>} Returns a Promise that on success will be resolved with an object with a bytesWritten property identifying the number of bytes written, and a buffer property that is data written.
   * @memberof I2csensor
   */
  _writeI2cBlock(command, buffer) {
    const cmd = this._validatedCommand(command);
    const buff = this._validatedBuffer(buffer);
    const { length } = buff;
    debug('writeI2cBlock. address: %h command: %h length: %h', this._address, cmd, length);
    return PromiseRetry((retry, number) => {
      debug('writeI2cBlock. attempt: %d', number);
      return this._bus.writeI2cBlock(this._address, cmd, length, buff).catch((err) => {
        debug('writeI2cBlock. error: %o', err);
        retry();
      });
    }, this._opts.retryOptions);
  }

  /**
   * Close the bus if created by sensor
   *
   * @returns {Promise<boolean>} Returns a Promise that will be resolved to true if bus created by sensor and closed successfully, or resolve to false if bus not created by sensor.
   * @memberof I2csensor
   */
  _closeI2c() {
    debug('closeI2c');
    return new Promise((resolve, reject) => {
      if (_.isNil(this._busnumber)) {
        resolve(false);
        return;
      }

      this._bus
        .close()
        .then(() => {
          resolve(true);
        })
        .catch((err) => {
          reject(err);
        });
    });
  }
}

module.exports = {
  I2csensor,
};


src/lib/i2csensor.defaults.js
/**
 * @typedef i2csensorDefaults
 * @type {object}
 * @property {object} retryOptions The promise-retry options {@link https://www.npmjs.com/package/promise-retry} for I2C communications
 * @property {number} retryOptions.retries=10 The maximum amount of times to retry the operation.
 * @property {number} retryOptions.factor=2 The exponential factor to use.
 * @property {number} retryOptions.minTimeout=1000 The number of milliseconds before starting the first retry.
 * @property {number} retryOptions.maxTimeout=Infinity The maximum number of milliseconds between two retries.
 * @property {boolean} retryOptions.randomize=false Randomizes the timeouts by multiplying with a factor between 1 to 2.
 */
const i2csensorDefaults = {
  retryOptions: {
    retries: 10,
    factor: 2,
    minTimeout: 1000,
    maxTimeout: Infinity,
    randomize: false,
  },
};

module.exports = {
  i2csensorDefaults,
};


src/bmp180sensor.js
const debug = require('debug')('Bmp180sensor');
const _ = require('lodash');
const delay = require('delay');

const { I2csensor } = require('./lib/i2csensor');
const { oversamplingSettings, registers } = require('./bmp180sensor.constants');

const { bmp180sensorDefaults } = require('./bmp180sensor.defaults');

/**
 * BMP180 Pressure and Temperature Sensor
 *
 * @class Bmp180sensor
 * @extends {I2csensor}
 */
class Bmp180sensor extends I2csensor {
  /**
   *Creates an instance of Bmp180sensor.
   * @param {object|number} bus An i2c-bus Bus class or the number for the bus
   * @param {number} [address=0x77] The I2C address
   * @param {object} options The configuration options
   * @param {object} [options.retryOptions] The promise-retry options {@link https://www.npmjs.com/package/promise-retry} for I2C communications
   * @param {number} [options.retryOptions.retries=10] The maximum amount of times to retry the operation.
   * @param {number} [options.retryOptions.factor=2] The exponential factor to use.
   * @param {number} [options.retryOptions.minTimeout=1000] The number of milliseconds before starting the first retry.
   * @param {number} [options.retryOptions.maxTimeout=Infinity] The maximum number of milliseconds between two retries.
   * @param {boolean} [options.retryOptions.randomize=false] Randomizes the timeouts by multiplying with a factor between 1 to 2.
   * @param {object} [options.oss] The oversampling setting
   * @param {number} options.oss.mode=1 The oversampling setting STANDARD mode
   * @param {number} options.oss.command=0x74 The oversampling setting STANDARD command
   * @param {number} options.oss.delay=8 The oversampling setting STANDARD delay in ms
   * @memberof Bmp180sensor
   */
  constructor(bus, address = 0x77, options) {
    const opts = _.merge({}, bmp180sensorDefaults, options);
    super('BMP180', bus, address, opts);
    this._registers = registers;
    this._calibrationdata = null;
    this._initialized = false;
    debug('Bmp180sensor constructor.');
  }

  /**
   * Calculate the B5 value used to convert raw values to compensated
   * temperature and pressure
   *
   * @param {*} ut Uncompensated temperature value
   * @returns {number} The B5 value used to convert raw values to compensated
   * @memberof Bmp180sensor
   *
   */
  _computeB5(ut) {
    if (!this.isCalibrated) {
      throw new Error('calibrationdata is invalid');
    }

    if (!Number.isInteger(ut)) {
      throw new TypeError(`Invalid Uncompensated Temperature ${ut}`);
    }

    const x1 = Math.floor(((ut - this._calibrationdata.ac6) * this._calibrationdata.ac5) / 2 ** 15);
    const x2 = (this._calibrationdata.mc * 2 ** 11) / (x1 + this._calibrationdata.md);
    const b5 = x1 + x2;
    return b5;
  }

  /**
   * Calculate the compensadate temperature in Celsius
   *
   * @param {number} ut Uncompensated temperature value
   * @returns {number} Temperature in 1 deg C units
   * @memberof Bmp180sensor
   */
  _convertUncompensatedTemperature(ut) {
    if (!Number.isInteger(ut)) {
      throw new TypeError(`Invalid Uncompensated Temperature ${ut}`);
    }
    const b5 = this._computeB5(ut);
    const temperature = Math.floor((b5 + 8) / 2 ** 4);
    return temperature / 10;
  }

  /**
   * Calculate the compensated pressure in Pascal
   *
   * @param {number} ut Uncompensated temperature value
   * @param {number} up Uncompensated pressure value
   * @returns {number} Pressure in Pa
   * @memberof Bmp180sensor
   */
  _convertUncompensatedPressure(ut, up) {
    if (!Number.isInteger(ut)) {
      throw new TypeError(`Invalid Uncompensated Temperature ${ut}`);
    }

    if (!Number.isInteger(up)) {
      throw new TypeError(`Invalid Uncompensated Pressure ${up}`);
    }
    const b5 = this._computeB5(ut);

    /* Pressure compensation */
    const b6 = b5 - 4000;
    let x1 = Math.floor((this._calibrationdata.b2 * Math.floor((b6 * b6) / 2 ** 12)) / 2 ** 11);
    let x2 = Math.floor((this._calibrationdata.ac2 * b6) / 2 ** 11);
    let x3 = x1 + x2;
    const b3 = Math.floor(
      ((this._calibrationdata.ac1 * 4 + x3) * 2 ** this._opts.oss.mode + 2) / 2 ** 2
    );
    x1 = Math.floor((this._calibrationdata.ac3 * b6) / 2 ** 13);
    x2 = Math.floor(Math.floor(this._calibrationdata.b1 * ((b6 * b6) / 2 ** 12)) / 2 ** 16);
    x3 = Math.floor((x1 + x2 + 2) / 2 ** 2);
    const b4 = Math.floor((this._calibrationdata.ac4 * (x3 + 32768)) / 2 ** 15);
    const b7 = (up - b3) * Math.floor(50000 / 2 ** this._opts.oss.mode);
    let p;
    if (b7 < 0x80000000) {
      p = (b7 * 2) / b4;
    } else {
      p = (b7 / b4) * 2;
    }
    x1 = Math.floor(p / 2 ** 8) * Math.floor(p / 2 ** 8);
    x1 = Math.floor((x1 * 3038) / 2 ** 16);
    x2 = Math.floor((-7357 * p) / 2 ** 16);
    const pressure = p + Math.floor((x1 + x2 + 3791) / 2 ** 4);
    return pressure;
  }

  /**
   * Read the uncompensated pressure
   *
   * @throws Will throw an error if invalid response buffer returned
   * @returns {Promise<number>} Returns a Promise that on success will be resolved to the uncompensated pressure (up)
   * @memberof Bmp180sensor
   */
  _getUncompensatedPressure() {
    debug('_getUncompensatedPressure');
    return this._writeI2cByte(this._registers.control, this._opts.oss.command)
      .then(() => {
        return delay(this._opts.oss.delay);
      })
      .then(() => {
        return this._readI2cBlock(this._registers.results, Buffer.alloc(3));
      })
      .then((response) => {
        const responseBuffer = this._validatedBuffer(response.buffer);
        if (responseBuffer.length < 3) {
          throw new Error('Results data not returned');
        }
        const MSB = responseBuffer.readUInt8(0);
        const LSB = responseBuffer.readUInt8(1);
        const XLSB = responseBuffer.readUInt8(2);
        const up = Math.floor(
          (MSB * 2 ** 16 + LSB * 2 ** 8 + XLSB) / 2 ** (8 - this._opts.oss.mode)
        );
        debug('_getUncompensatedPressure (up) %d', up);
        return up;
      });
  }

  /**
   * Read the uncompensated temperature
   *
   * @throws Will throw an error if invalid response buffer returned
   * @returns {Promise<number>} Returns a Promise that on success will be resolved to the uncompensated temperature (ut)
   * @memberof Bmp180sensor
   */
  _getUncompensatedTemperature() {
    debug('_getUncompensatedTemperature');
    return this._writeI2cByte(this._registers.control, oversamplingSettings.temperature.command)
      .then(() => {
        return delay(oversamplingSettings.temperature.delay);
      })
      .then(() => {
        return this._readI2cBlock(this._registers.results, Buffer.alloc(2));
      })
      .then((response) => {
        const responseBuffer = this._validatedBuffer(response.buffer);
        if (responseBuffer.length < 2) {
          throw new Error('Results data not returned');
        }
        const MSB = responseBuffer.readUInt8(0);
        const LSB = responseBuffer.readUInt8(1);
        const ut = MSB * 2 ** 8 + LSB;
        debug('_getUncompensatedTemperature (ut) %d', ut);
        return ut;
      });
  }

  /**
   * Check that sensor is calibrated
   *
   * @readonly
   * @memberof Bmp180sensor
   */
  get isCalibrated() {
    return !_.isNil(this._calibrationdata);
  }

  /**
   * Check that sensor is calibrated
   *
   * @readonly
   * @memberof Bmp180sensor
   */
  get isInitialized() {
    return this._initialized;
  }

  /**
   * Get the oversampling setting value
   *
   * @readonly
   * @memberof Bmp180sensor
   */
  get oss() {
    return this._opts.oss;
  }

  /**
   * Set the oversampling setting value
   *
   * @memberof Bmp180sensor
   */
  set oss(value) {
    this._opts.oss = value;
  }

  /**
   * Validates the chip id is correct for a BMP180 device
   *
   * @returns {Promise<boolean>} Returns a Promise that on success will be resolved to true
   * @throws Will throw if chip id is invalid.
   * @memberof Bmp180sensor
   */
  isValidChipId() {
    debug('checkChipId');
    return this._readI2cByte(this._registers.chipid).then((response) => {
      if (response !== 0x55) {
        throw new Error(
          `Expected BMP180 chip_id to be 0x55, got chip id 0x${response.toString(16)}.`
        );
      }
      return true;
    });
  }

  /**
   * Perform a Soft Reset this will perform the same sequence as power on reset.
   *
   * @returns {Promise<boolean>} Returns a Promise that on success will be resolved to true
   * @memberof Bmp180sensor
   *
   */
  softreset() {
    debug('softreset');
    return this._writeI2cByte(this._registers.softreset, 0xb6).then(() => {
      return true;
    });
  }

  /**
   * Retrieves chip calibration data, and stores in this.calibrationData.
   *
   * @returns {Promise<boolean>} Returns a Promise that on success will be resolved to true
   * @throws Will throw if calibration data returned is invalid.
   * @memberof Bmp180sensor
   *
   */
  calibrate() {
    debug('calibrate');
    const calLength = 22;
    return this._readI2cBlock(this._registers.calibration, Buffer.alloc(calLength)).then(
      (response) => {
        const responseBuffer = this._validatedBuffer(response.buffer);
        if (responseBuffer.length < calLength) {
          throw new Error('Calibration data not returned');
        }

        this._calibrationdata = Object.freeze({
          ac1: responseBuffer.readInt16BE(0),
          ac2: responseBuffer.readInt16BE(2),
          ac3: responseBuffer.readInt16BE(4),
          ac4: responseBuffer.readUInt16BE(6),
          ac5: responseBuffer.readUInt16BE(8),
          ac6: responseBuffer.readUInt16BE(10),
          b1: responseBuffer.readInt16BE(12),
          b2: responseBuffer.readInt16BE(14),
          mb: responseBuffer.readInt16BE(16),
          mc: responseBuffer.readInt16BE(18),
          md: responseBuffer.readInt16BE(20),
        });

        debug('calibrate. data %o', this._calibrationdata);
        return true;
      }
    );
  }

  /**
   * Initializes the sensor, calling:
   * isValidChipId()
   * softReset()
   * calibrate()
   *
   * @returns {Promise<boolean>} Returns a Promise that on success will be resolved to true
   * @memberof Bmp180sensor
   *
   */
  initialize() {
    debug('initialize');
    this._initialized = false;
    return this.isValidChipId()
      .then(() => {
        return this.softreset();
      })
      .then(() => {
        return this.calibrate();
      })
      .then(() => {
        this._initialized = true;
        return this._initialized;
      });
  }

  /**
   * Closes the I2c bus
   *
   * @returns {Promise} Returns a Promise that will be resolved with no arguments
   * @memberof Bmp180sensor
   *
   */
  close() {
    debug('close');
    return this._closeI2c();
  }

  /**
   * Current temperature in Celsius
   *
   * @returns {Promise<number>} Returns a Promise that on success will be resolved to the current temperature in Celsius
   * @memberof Bmp180sensor
   *
   */
  temperature() {
    debug('temperature');
    return new Promise((resolve, reject) => {
      this._getUncompensatedTemperature().then((ut) => {
        try {
          const value = this._convertUncompensatedTemperature(ut);
          debug('temperature. %d C', value);
          resolve(value);
        } catch (err) {
          reject(err);
        }
      });
    });
  }

  /**
   * Current pressure in Pascals
   *
   * @returns {Promise<number>} Returns a Promise that on success will be resolved to the current pressure in Pascals
   * @memberof Bmp180sensor
   *
   */
  pressure() {
    debug('pressure');
    return new Promise((resolve, reject) => {
      this._getUncompensatedTemperature().then((ut) => {
        this._getUncompensatedPressure().then((up) => {
          try {
            const value = this._convertUncompensatedPressure(ut, up);
            debug('pressure. %d Pa', value);
            resolve(value);
          } catch (err) {
            reject(err);
          }
        });
      });
    });
  }
}

module.exports = {
  Bmp180sensor,
};
